### PR TITLE
Refactor C linters for better extensibility/reuse

### PR DESCRIPTION
**Pull request by MichaelChirico**
Most importantly, we can run `readLines()` + preprocess the file only once, then pass that output to each linter.

### CODE REVIEW

**Comment by MichaelChirico** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/Rdatatable/data.table/6346?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#6347** <a href="https://app.graphite.dev/github/pr/Rdatatable/data.table/6347?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6346** <a href="https://app.graphite.dev/github/pr/Rdatatable/data.table/6346?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#6345** <a href="https://app.graphite.dev/github/pr/Rdatatable/data.table/6345?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @MichaelChirico and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: .ci/linters/c/alloc_linter.R [hunk 1]
```diff
@@ -2,8 +2,8 @@
 # More specifically, given that this is an AST-ignorant checker,
 #   1. Find groups of malloc()/calloc() calls
 #   2. Check the next line for a check like 'if (!x || !y)'
-alloc_linter = function(c_file) {
-  lines <- readLines(c_file)
+alloc_linter = function(c_obj) {
+  lines <- c_obj$lines
```
Comment (ben-schwen) [inline .ci/linters/c/alloc_linter.R:6]: I know its not directly `data.table` but still feels strange that we do not use `=` for assignment in this file (style linting the linter files)

**Review by ben-schwen — APPROVED** [review]
LGTM (besides `<-` and `=` mix)

**Comment by MichaelChirico** [comment]
### Merge activity

* **Aug 19, 12:05 PM EDT**: @MichaelChirico started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/Rdatatable/data.table/6346).

**Pull request merged**